### PR TITLE
fix(sidenav): sidenavmenu stays open when in rail

### LINKED ISSUE #3527: [UI Shell] [Rails] - Submenu should collapse when rail is closed
##  What package(s) are you using?

- [ ] `carbon-components`
- [x] `carbon-components-react`

## Detailed description

> Describe in detail the issue you're having.

If rails is collapsed but the submenu is still expanded, you see additional whitespace which is essentially the submenu items. Additionally, if one of these submenu items is focused or selected, you see the active state but it looks off since you can't see the submenu items in collapsed rails.

rails-bug

> Is this issue related to a specific component?

UI Shell rails variation

> What did you expect to happen? What happened instead? What would you like to
> see changed?

The submenu should also collapse so that there's no whitespace from the submenu items.

rails-bug2

## Steps to reproduce the issue

1. Step one: Expand rails
2. Step two: Open a submenu and select a submenu item
3. Step three: Collapse rails
4. Step four: Notice submenu bug

### PR DESCRIPTION
**Pull request by matthew-chirgwin**
Closes #3527

The `SideNavMenu` component when used in a `SideNav` (using `isRail` property) maintains its space for its child items when the rail is collapsed if selected. This change adds a property to determine if the `SideNav` is open, and in the case of the `SideNavMenu`, change state if it was open/closed when the rail is open/closed.

#### Changelog

**Changed**

- Added `isSideNavExpanded` property to child components of the SideNav
- Added `gDSFP` logic to `SideNavMenu` to detect and handle this case
- Added tests for the above

#### Testing / Reviewing

Unit tests have been added to check on the `gDSFP` logic to `SideNavMenu` to make sure this issue could not return . If more testing is required please let me know and Ill update the PR.

### CODE REVIEW

**Review by jnm2377 — APPROVED** [review]
LTGM! 🎉

**Comment by tw15egan** [comment]
Seems like this PR has broken functionality for other versions of the UI Shell: https://deploy-preview-3626--carbon-components-react.netlify.com/?path=/story/ui-shell--header-base-w-navigation-actions-and-sidenav

I'm not able to open any of the left nav menu's

**Comment by matthew-chirgwin** [comment]
Good spot @tw15egan I will look into that now

**Comment by tw15egan** [comment]
And since this was the main bug blocking the Rail version from being displayed, should we go ahead and add a story for the Rails version in now that it seems to be working properly? Or do we want to wait until keyboard functionality is worked in as well (https://github.com/carbon-design-system/carbon/issues/3529)?

cc @jnm2377 and @aledavila for your opinions

**Comment by aledavila** [comment]
@tw15egan yea I had suggested we display now if it does work. Since the other one is an enhancement

**Comment by matthew-chirgwin** [comment]
Hi all,

As requested, I have added the story in to the storybook. I don’t know if I need to include the build of the storybook or not in the PR, but have not for now. If it is needed, let me know.

As for the issue @tw15egan found, I have guarded the passing of expansion state so it only does this if `isRail` is passed. From the testing I have done, this maintains the existing capability, but also fixes this issue.

Just one note on why this happened originally too. I was expecting the `expanded` prop (if provided) to dictate the rendering of the `SideNav` (ie it being expanded or not). Even when false, it still renders the navigation section, and its items (this is the case without my change and with it). With the change I have made, the`SideNavMenu` now only expands/responds if the `SideNav` has an expanded prop/internal state value of `true`, hence the change in behaviour. Given the prop description, I’m not sure if the `SideNav` is doing what it should, or if the prop description need updating.

**Comment by aledavila** [comment]
Hi thanks @matthew-chirgwin 

Its looking good so far. But there is still an issue the menu now doesn't pin open on click of the header menu. There are 2 ways that the rail should open. On hover of the side nav or on click of the header menu and it should remain open. I believe it was working before.

**Comment by tw15egan** [comment]
The story looks great, and it seems like you've got open on focus behavior sorted. Awesome! I see that the other stories are now working as intended. 

I'm also noticing the issue @aledavila pointed out. 

In addition, would it be possible to add the active state to the closed menu item, like so:



Looking good so far, and thanks for all the work you've put into this 👍

**Comment by matthew-chirgwin** [comment]
Hey @aledavila @tw15egan . Thanks for the feedback. I've added the required props/callbacks in to allow the rails story to work as you expected (I had removed them for local testing and did not put them back). That has just been added to the branch now ready for review.

I also had a (quick) look at adding the active state when collapsed. With the change as I have it, an active menu will keep the styling even if I interact with another item:

Screenshot 2019-08-02 at 20 48 26

Is it a must have for this issue, or worth a follow on issue?

**Comment by aledavila** [comment]
Thanks for the update. It seems that now the submenus don't open on hover. But if you click them open and then click the header menu the ones you clicked to open are open. So it seems to be passing in the open prop but it doesn't actually open on hover.

**Comment by vpicone** [comment]
@asudoh @joshblack  Is this really the prefered method instead of using context? I don't see how mapping over the children is any more explicit than context, the only difference is it's less flexible and transparent for developers.

**Comment by asudoh** [comment]
Propagate-prop-to-child pattern is one we have employed since the inception of the codebase. It has "unintentional override of props" problem and I personally have considered of switching it to render-props, but I haven't figured out a pattern that works for every affected components (e.g. tabs). That said, using propagate-prop-to-child pattern here is fine to me.

**Comment by vpicone** [comment]
@asudoh a major issue you had with context is that the passing of props wasn't explicit to developers. How is this pattern any different?

**Comment by asudoh** [comment]
Propagate-prop-to-child doesn't break the deterministic behavior of components upon props.

**Comment by joshblack** [comment]
If I understand things correctly, what @asudoh is referring to is that using `cloneElement` means that these components will work without the provider (let me know if I'm misunderstanding 👍)

In your proposal, Vince, were you suggesting having a default provider embedded in our components or would it noop with the default arguments passed into `createContext`?

**Comment by matthew-chirgwin** [comment]
Hi all - just as an FYI, I am on vacation this week, which means I won’t be able to make any immediate changes to this PR. I don’t object to updating to use context if that is the preference - I went with props as I started the change from the `SideNavMenu`, and didn’t realise how many levels of prop passing were required. I can update if needed on Monday. Re your last comment @aledavila ,  plumbing in the render prop for the header button (as requested) into the `SideNav` expanded prop means the SideNav become controlled, and thus looses the rails mouse over/out behaviour. Please correct me if I have that wrong. Is there another method/prop you were expecting to enable both?

**Comment by aledavila** [comment]
@matthew-chirgwin Sorry what I meant is that when you hover over the rails and try to click a submenu to open it. It won't open.

**Comment by asudoh** [comment]
> Propagate-prop-to-child doesn't break the deterministic behavior of components upon props.

Just to clarify more on that - Regular components tend to be guaranteed that how it works is deterministic upon props, which means, unless props change, the behavior doesn't change. Propagate-prop-to-child pattern changes the behavior of components as they are put as children, but still the parent is explicit about changing props.

**Comment by joshblack** [comment]
@asudoh I think what @vpicone proposes when bringing up context is a similar relation, but instead of exposing the behavior through an external-facing API it uses an internal API. For example:

```jsx
function ExplicitChild({ value }) {
  // ...
}

function InternalChild() {
  const value = useContext(CustomContextProvider);

  // ...
}
```

Providing the value as props has the limitation of being associated to a particular owner tree, whereas the internal API could open up the possibility to make things more flexible in terms of the owner tree. This is to say that we are tied to an explicit hierarchy such as:

```jsx
<ExplicitParent>
  <ExplicitChild />
</ExplicitParent>
```

Such that `ExplicitParent` uses `cloneElement` to provide an internal value to `ExplicitChild`. As a result, we run into the problem where a custom wrapper breaks this flow unless they know which exact properties to expose to its child:

```jsx
<ExplicitParent>
  <CustomWrapper>
    <ExplicitChild />
  </CustomWrapper>
</ExplicitParent>
```

`CustomWrapper` in this case would have to know about these internal state values in order to pass them along to `ExplicitChild`. Ultimately, for a consumer who is not concerned with managing this state value, this is seen as a drawback to this API design.

While some may refer to this as prop drilling, there is an extra layer of complexity here in that this is an internal state value that we use internally instead of a prop designed for external use. For example, we may want to be able to change internal state shape without it being considered a change on the consumer/caller's side.

With the context-based API, we're able to bypass these limitations while still having a strict contract on where the value comes from (a parent provider). Components could be written such that they expose this provider at the top-level:

```jsx
const InternalContext = React.createContext({ value: null });

function InternalParent({ children }) {
  const value = useState(someValue);
  return (
    <InternalContext.Provider value={value}>
      {children}
    </InternalContext.Provider>
  );
}

function InternalChild() {
  const value = useContext(InternalContext);
  // ...
}
```

And now we don't have to necessarily worry about the owner tree anymore, as context can go through any layer and work as-intended:

```jsx
<InternalParent>
  <CustomWrapper>
    <InternalChild />
  </CustomWrapper>
</InternalParent>
```

At the end of the day, this helps to encapsulate internal state in our components such that they do not leak to consumers, this has the following effects:

- Consumers no longer need to manually manage internal state values for wrappers
- Changing state shape no longer constitutes a change to a component's external API (props)
- An owner/child API contract is still maintained through the explicitly created context, `InternalContext`

While this definitely can be associated with some drawbacks, hope this can help to explain why this concept is being brought up (as I understand it) 👍

**Comment by vpicone** [comment]
@joshblack woah, that was amazing. Way better than I could have explained it thanks! I think the last three points summed it up nicely. In my implementation, explicit props passed by developer takes precedence over the context value (which, yeah, would have a default value provided by createContext). Each context consumer can be controlled but falls back to using the context value.

**Comment by asudoh** [comment]
If `<InternalChild>` something that we let consumers render, it'll be fine. OTOH if `<InternalChild>` is something our component renders internally, it makes that our component non-deterministic.

**Comment by joshblack** [comment]
@asudoh could you explain more? Initially I thought you were describing the limitation here that `InternalChild` could then not be used outside of `InternalParent`. In this case, they would need to wire things up much like they need to with props but now with a context provider so it seems like roughly the same situation just with a different call syntax.

**Comment by matthew-chirgwin** [comment]
Hi guys,

From the discussion, would the preference be I swap to using context (pending any other discussion with @asudoh )? I wasnt a fan of the cloning/prop passing in my first pass at this, but also didn't want to make context a requirement for usage. If the desire is for a context based solution, please let me know and I will get on making the changes required.

@aledavila good spot. The issue now is the rail expands on hover (which is fine) via css, but from a react side, the component is controlled, hence when you clicked on the SideNavMenu it did nothing as it is being told it is closed/should not change state. I'll have a think about options, but is changing the CSS to not use a hover class going to be an issue? I imagine it may be for the other versions of the SideNav!

**Comment by asudoh** [comment]
@matthew-chirgwin I don't want our components to depend on React context, either. As I stated above, if React dependency is something application (instead of our components) makes, it'll be fine. Though I'm not a big fan of propagating-props-to-child pattern either, I'm OK with it, too.

**Comment by joshblack** [comment]
@matthew-chirgwin sorry about the extra chatter on the PR! Definitely caught us in a conversation we've been having internally on this topic 😄 

@asudoh it would be helpful if you could address the above notes either here or in our Slack channel 👍 As it stands, would be great to get consensus and understanding across all the different perspectives.

**Comment by matthew-chirgwin** [comment]
Hi all,

I have just updated my branch witha fix for the issue @aledavila spotted. No problem with the discussion @joshblack / @asudoh , if there are changes you would like made off the back of that discussion, please let me know.

**Review by asudoh — APPROVED** [review]
The latest code LGTM 👍 - Thanks @matthew-chirgwin!

**Comment by matthew-chirgwin** [comment]
Hi all, I have also had a look at issue #3529 , and incorparated a fix for that into this PR (see commit 0d177f720e34f1147809cdaa66560cd7fdb77feb ). If you have any feedback please let me know!

File: packages/react/src/components/UIShell/SideNav.js [hunk 1]
```diff
@@ -78,9 +99,9 @@ const SideNav = React.forwardRef(function SideNav(props, ref) {
         {...accessibilityLabel}
         onFocus={event => handleToggle(event, true)}
         onBlur={event => handleToggle(event, false)}
-        onMouseEnter={() => handleToggle(true)}
-        onMouseLeave={() => handleToggle(false)}>
-        {children}
+        onMouseEnter={() => handleToggle({}, true)}
```
Comment (asudoh) [inline packages/react/src/components/UIShell/SideNav.js]: Any reason not passing along the event?

**Review by asudoh — COMMENTED** [review]
The latest update makes sense, just one thing - Thanks @matthew-chirgwin!

File: packages/react/src/components/UIShell/SideNav.js (hunk 1 above)
Comment (matthew-chirgwin) [reply packages/react/src/components/UIShell/SideNav.js]: Hi @asudoh good point - as it was returning `true`/`false` before, it should continue to do so. I will change this to `onMouseEnter={() => handleToggle(true, true)}` and `onMouseLeave={() => handleToggle(false, false)}`

**Review by tw15egan — APPROVED** [review]
This looks really great! Thanks for making all these changes/improvements👍

**Comment by designertyler** [comment]
From a visual perspective, this looks good. The only thing I wanted to double check is the active state when a menu item is selected. I wasn't able to verify that in the demo and wasn't sure if that's a limitation of the demo or overlooked.

I was hoping to see this state
image

File: packages/react/src/components/UIShell/SideNav.js (hunk 1 above)
Comment (asudoh) [reply packages/react/src/components/UIShell/SideNav.js]: > as it was returning `true`/`false` before, it should continue to do so.

Right. An interesting part is (as you may have noticed) that the old code didn't have one-arg interface. And other callers of `handleToggle()` passes along the raw event. So I was wondering if we could do:

```suggestion
        onMouseEnter={event => handleToggle(event, true)}
```

**Comment by jnm2377** [comment]
wrt @designertyler comment above. It looks like the active `SideNavMenu` is just missing `isActive={true}`. 

```
<SideNavMenu
   renderIcon={Fade16}
   title="Category title"
   isActive={true}>
```

**Comment by jnm2377** [comment]
rails
Working as expected when you add the missing prop 👍

**Comment by matthew-chirgwin** [comment]
Hi @designertyler @jnm2377 /all - as it turns out, the was a small bug in the `SideNavMenu` around adding the active class (which can also be driven by the isActive prop). I have fixed this in my latest commit (so it does actually check it's children/their props), meaning it now has the behaviour you requested, and you don't have to explicitly set the prop on the menu if you set the active state on the child. I also added tests to cover this. I hope that is ok!

**Review by designertyler — APPROVED** [review]
Looks good to me!

File: packages/react/src/components/UIShell/SideNav.js [hunk 2]
```diff
@@ -78,9 +99,9 @@ const SideNav = React.forwardRef(function SideNav(props, ref) {
         {...accessibilityLabel}
         onFocus={event => handleToggle(event, true)}
         onBlur={event => handleToggle(event, false)}
-        onMouseEnter={() => handleToggle(true)}
-        onMouseLeave={() => handleToggle(false)}>
-        {children}
+        onMouseEnter={() => handleToggle(true, true)}
```
Comment (asudoh) [inline packages/react/src/components/UIShell/SideNav.js:102]:
```suggestion
        onMouseEnter={event => handleToggle(event, true)}
```

File: packages/react/src/components/UIShell/SideNav.js [hunk 3]
```diff
@@ -78,9 +99,9 @@ const SideNav = React.forwardRef(function SideNav(props, ref) {
...
+        onMouseLeave={() => handleToggle(false, false)}>
```
Comment (asudoh) [inline packages/react/src/components/UIShell/SideNav.js:103]:
```suggestion
        onMouseLeave={event => handleToggle(event, false)}>
```

**Review by asudoh — COMMENTED** [review]
Just small things to ensure that it's no lost in the pile of conversation - Thank you for all the updates @matthew-chirgwin!

File: packages/react/src/components/UIShell/SideNav.js (hunk 2 above)
Comment (matthew-chirgwin) [reply packages/react/src/components/UIShell/SideNav.js:102]: @asudoh - before my changes, `handleToggle` was called here with just true as the only argument. I can change it to whatever the event is as you suggest, but if someone downstream passes the `onToggle` property to the `SideNav` and expects the bool value, would that not be a breaking change for them?

File: packages/react/src/components/UIShell/SideNav.js (hunk 3 above)
Comment (matthew-chirgwin) [reply packages/react/src/components/UIShell/SideNav.js:103]: @asudoh - same here as my comment above

File: packages/react/src/components/UIShell/SideNav.js (hunk 2 above)
Comment (asudoh) [reply packages/react/src/components/UIShell/SideNav.js:102]: Thank you for putting your thoughts on this and elaborating it @matthew-chirgwin! I now see your point (I missed it, tricked by argument name), and I now think the best is keeping `handleToggle()` call intact - Meaning, call with one arg.

Comment (matthew-chirgwin) [reply packages/react/src/components/UIShell/SideNav.js:102]: Hi @asudoh - the only issue with moving to one argument is the effect that would then have on how I changed `handleToggle` to handle this case (lines 48-50). That change could be pushed into its own function, but would that add any value?

**Comment by jnm2377** [comment]
Update for reviewers: we decided we'll be going with context based components for UI shell. However, given that we need to update more than just the rails functionality to use context (this is a bigger effort than the scope of this PR), we are merging this in as-is and will update to context in a separate PR.

**Comment by jnm2377** [comment]
Issue for context: https://github.com/carbon-design-system/carbon/issues/3797

File: packages/react/src/components/UIShell/SideNav.js (hunk 2 above)
Comment (asudoh) [reply packages/react/src/components/UIShell/SideNav.js:102]: Just leaving a note, though this is merged already; I think the best approach is making `handleToggle()` obvious that the first arg is optional and and the second arg becomes the first arg when the caller doesn't specify the first arg.

**Pull request merged**